pkg/pillar/volumehandlers/csihandler.go

```diff
@@ -0,0 +1,331 @@
+			handler.log.Error(errStr)
+			return "", errors.New(errStr)
+		}
+	}
+
+	handler.log.Functionf("CreateVolume(%s) DONE", pvcName)
+	return pvcName, nil
+}
+
+func (handler *volumeHandlerCSI) DestroyVolume() (string, error) {
+	pvcName := handler.status.Key()
+	handler.log.Noticef("DestroyVolume called for PVC %s", pvcName)
+	err := kubeapi.DeletePVC(pvcName, handler.log)
+	if err != nil {
+		// Its OK if not found since PVC might have been deleted already
+		if kerr.IsNotFound(err) {
+			handler.log.Noticef("PVC %s not found, might have been deleted", pvcName)
+			return pvcName, nil
+		} else {
+			return pvcName, err
+		}
+	}
+	return pvcName, nil
+}
+
+func (handler *volumeHandlerCSI) Populate() (bool, error) {
+	handler.log.Noticef("Populate called for PVC %s", handler.status.Key())
+	_, err := kubeapi.FindPVC(handler.status.Key(), handler.log)
+	if err != nil {
+		// Its OK if not found since PVC might not be created yet.
+		if kerr.IsNotFound(err) {
+			handler.log.Noticef("PVC %s not found", handler.status.Key())
+			return false, nil
+		} else {
+			return false, err
+		}
+	}
+	return true, nil
+}
+
+// Copied from zvolhandler.go
+func (handler *volumeHandlerCSI) getVolumeFilePath() (string, error) {
+	puller := registry.Puller{
+		Image: handler.status.ReferenceName,
+	}
+	ctrdCtx, done := handler.volumeManager.GetCasClient().CtrNewUserServicesCtx()
+	defer done()
+
+	resolver, err := handler.volumeManager.GetCasClient().Resolver(ctrdCtx)
+	if err != nil {
+		errStr := fmt.Sprintf("error getting CAS resolver: %v", err)
+		handler.log.Error(errStr)
+		return "", errors.New(errStr)
+	}
+	pathToFile := ""
+	_, i, err := puller.Config(true, os.Stderr, resolver)
+	if err != nil {
+		errStr := fmt.Sprintf("error Config for ref %s: %v", handler.status.ReferenceName, err)
+		handler.log.Error(errStr)
+		return "", errors.New(errStr)
+	}
+	if len(i.RootFS.DiffIDs) > 0 {
+		// FIXME we expects root in the first layer for now
+		b := i.RootFS.DiffIDs[0]
+		// FIXME we need the proper way to extract file from content dir of containerd
+		pathToFile = filepath.Join(types.ContainerdContentDir, "blobs", b.Algorithm().String(), b.Encoded())
+	}
+
+	if pathToFile == "" {
+		errStr := fmt.Sprintf("no blobs to convert found for ref %s", handler.status.ReferenceName)
+		handler.log.Error(errStr)
+		return "", errors.New(errStr)
+	}
+	return pathToFile, nil
+}
+
+// Creates a dest qcow2 file from the srcLocation
+func (handler *volumeHandlerCSI) CopyImgDirToRawImg(ctx context.Context, log *base.LogObject, srcLocation string, destFile string) error {
+
+	args := []string{srcLocation, destFile}
+
+	log.Noticef("%s args %v", imageToQcowScript, args)
+
+	output, err := base.Exec(log, imageToQcowScript, args...).WithContext(ctx).WithUnlimitedTimeout(432000 * time.Second).CombinedOutput()
+
+	if err != nil {
+		errStr := fmt.Sprintf("CopyImgDirToRawImg: Failed to create raw image file  %s: %v", output, err)
```

Comment:
```suggestion
		errStr := fmt.Errorf("CopyImgDirToRawImg: Failed to create raw image file  %s: %v", output, err)
```

Reply:
If we apply this change, the var should be changed as well: `err`, not `errStr` and returned directly without calling `error.New()`...

Reply:
Done